peer/pendingheap/heap.go

```diff
@@ -166,6 +172,28 @@ func (ph *pendingHeap) pushPeer(ps *peerScore) {
 	heap.Push(ph, ps)
 }
 
+// pushPeerRandom must be called in the context of a lock.
+//
+// It inserts a peer randomly into the heap among equally scored peers. This is
+// expected to be called only by Add.
+//
+// This ensures that batches of peer updates are inserted randomly throughout
+// the heap, preventing hearding behavior that may be observed during batch
+// deployments.
+func (ph *pendingHeap) pushPeerRandom(ps *peerScore) {
+	ph.next++
+	ps.last = ph.next
+
+	random := ph.nextRand(len(ph.peers) + 1)
+	if random < len(ph.peers) {
+		randPeer := ph.peers[random]
+		ps.last, randPeer.last = randPeer.last, ps.last
+		heap.Fix(ph, random)
+	}
```

Comment:
I’m pretty sure there’s a way to phrase this that eliminates the `if` clause by cleverly reordering and making the swap an implied noop.